src/main/java/com/testingsyndicate/jms/responder/MessageHandler.java

```diff
@@ -0,0 +1,138 @@
+package com.testingsyndicate.jms.responder;
+
+import com.testingsyndicate.jms.responder.model.RequestInfo;
+import com.testingsyndicate.jms.responder.model.StubbedResponse;
+import com.testingsyndicate.jms.responder.repository.ResponseRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jms.*;
+import javax.jms.IllegalStateException;
+import java.util.Optional;
+
+final class MessageHandler implements Runnable {
+
+    private static final int RECEIVE_TIMEOUT = 2000;
+
+    private static final Logger LOG = LoggerFactory.getLogger(MessageHandler.class);
+
+    private final Session session;
+    private final String queueName;
+    private final ResponseRepository repository;;
+
+    MessageHandler(Session session, String queueName, ResponseRepository repository) {
+        this.session = session;
+        this.queueName = queueName;
+        this.repository = repository;
+    }
+
+    @Override
+    public void run() {
+        MessageConsumer consumer = null;
+
+        try {
+            Destination destination = session.createQueue(queueName);
+            consumer = session.createConsumer(destination);
+
+            while (true) {
+                try {
+                    LOG.info("Waiting for message on {} for {}ms", destination, RECEIVE_TIMEOUT);
+                    Message request = consumer.receive(RECEIVE_TIMEOUT);
+                    if (null != request) {
+                        onMessage(request);
+                    }
+                } catch (IllegalStateException ise) {
+                    LOG.info("Shutting down");
+                    break;
+                } catch (JMSException ex) {
+                    LOG.warn("Error in message loop, {}", ex.getMessage());
+                }
+            }
+
+        } catch (JMSException ex) {
+            LOG.error("Error in thread startup", ex);
+            throw new RuntimeException("Error on startup", ex);
+        } finally {
+            if (null != consumer) {
+                try {
+                    consumer.close();
+                } catch (JMSException e) {
+                    LOG.error("Could not close consumer", e);
+                }
+            }
+        }
+    }
+
+    void onMessage(Message message) {
+
+        LOG.info("Received message {}", message);
+
+        try {
+            if (message instanceof TextMessage) {
```

Comment:
What do you think of making this more 'guard'-like, so we can better associate the reason we're bombing out with the step we take after bombing out, as well as getting rid of the nesting?

```java
if (!(message instanceof TextMessage)) {
    LOG.info("Ignoring message (Not TextMessage, is {})", message.getClass());
    return;
}

TextMessage textMessage = (TextMessage) message;

// Find a stub
Queue queue = (Queue) textMessage.getJMSDestination();

RequestInfo requestInfo = RequestInfo.newBuilder()
        .withBody(textMessage.getText())
        .withQueueName(queue.getQueueName())
        .build();

LOG.trace("Looking for a match");
Optional<StubbedResponse> match = repository.findMatch(requestInfo);

if (!match.isPresent()) {
    LOG.warn("No match found, not sending a reply");
    return;
}

StubbedResponse response = match.get();
LOG.debug("Found match {}", response);
```

Reply:
Good idea, I don't know why I didn't do that because it's also more my style